### PR TITLE
ENT-3610: Put usage metric id in the product registry yaml

### PR DESCRIPTION
**Pull request by dmckoske**
The testing for this is mainly unit testing, but I'm open to suggestions.

### CODE REVIEW

**Review by mstead — CHANGES_REQUESTED** [review]
After chatting with @kahowell I believe there is a part missing in this PR. Kevin mentioned that it was implied that ENT-3610 would include hooking the new configuration bits into appropriate place.

We need to get the metric value from the ProductProfile and replace the hard coded value here:
https://github.com/RedHatInsights/rhsm-subscriptions/blob/45f2786f45ccd316f1fa42188945dd34b34067f4/src/main/java/org/candlepin/subscriptions/files/ProductProfile.java#L75

productProfile.getMetricId() is called from the MarketplacePayloadMapper. You will need to make the connection.

File: src/main/java/org/candlepin/subscriptions/files/MarketplaceMetrics.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Represents marketplace metric features, including swatch product IDs.
+ */
+@Getter
+@Setter
+public class MarketplaceMetrics {
```
[the remark this answers is not in the document]
Comment (dmckoske) [reply src/main/java/org/candlepin/subscriptions/files/MarketplaceMetrics.java:35]: Fair.  Done.

File: src/main/java/org/candlepin/subscriptions/files/MarketplaceMetrics.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    private java.lang.String metricId;
+    private java.lang.String uom;
+    private Set<String> swatchProductIds;
+
+
+    public MarketplaceMetrics() {
+        this.swatchProductIds = new HashSet<>();
+    }
+
+    public MarketplaceMetrics(java.lang.String metricId, java.lang.String uom, Set<String> swatchProductIds) {
```
[the remark this answers is not in the document]
Comment (dmckoske) [reply src/main/java/org/candlepin/subscriptions/files/MarketplaceMetrics.java:45]: Not sure how this happened.  Changed throughout.

File: src/main/java/org/candlepin/subscriptions/files/MarketplaceMetrics.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        MarketplaceMetrics that = (MarketplaceMetrics) o;
+        return metricId.equals(that.metricId) &&
+            uom.equals(that.uom) &&
+            swatchProductIds.equals(that.swatchProductIds);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(metricId, uom, swatchProductIds);
+    }
```
[the remark this answers is not in the document]
Comment (dmckoske) [reply src/main/java/org/candlepin/subscriptions/files/MarketplaceMetrics.java:69]: Fair.  Done.

File: src/main/java/org/candlepin/subscriptions/files/ProductUom.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+ */
+package org.candlepin.subscriptions.files;
+
+import lombok.Getter;
+import lombok.Setter;
+
+import java.util.Objects;
+
+/**
+ * A set object for marketplace metrics.
```
[the remark this answers is not in the document]
Comment (dmckoske) [reply src/main/java/org/candlepin/subscriptions/files/ProductUom.java:29]: Fair.  Done.

File: src/test/java/org/candlepin/subscriptions/files/ProductProfileRegistryTest.java [hunk 5]
```diff
@@ -69,10 +70,19 @@ void setUp() {
             makeRole("os-metrics", Set.of(OPENSHIFT_METRICS)),
             makeRole("ocp", Set.of(OPENSHIFT_DEDICATED_METRICS))
         ));
+
+        ProductProfile p5 = new ProductProfile("p5", Collections.emptySet(), HOURLY);
+
+        p5.setMarketplaceMetrics(Set.of(
+            makeMarketplaceMetrics("redhat.com:openshiftdedicated:cpu_hour",
+            Measurement.Uom.CORES.toString(), Set.of(OPENSHIFT_DEDICATED_METRICS.toString()))
```
[the remark this answers is not in the document]
Comment (dmckoske) [reply src/test/java/org/candlepin/subscriptions/files/ProductProfileRegistryTest.java:78]: Fair.  Done.

File: src/main/java/org/candlepin/subscriptions/files/ProductProfileRegistry.java [hunk 6]
```diff
@@ -197,4 +202,16 @@ public void validateGranularityCompatibility(ProductId productId, Granularity re
...
+    /**
+     * Look up a metric ID from its swatch profile ID and measurement UOM.
+     *
+     * @param swatchProfileId swatch profile ID
+     * @param uom measurement UOM
+     * @return the metric
+     */
+    public java.lang.String lookupMetricId(java.lang.String swatchProfileId, Measurement.Uom uom) {
+        return this.metricByProductAndUom.get(new ProductUom(swatchProfileId, uom.toString()));
+    }
```
[the remark this answers is not in the document]
Comment (dmckoske) [reply src/main/java/org/candlepin/subscriptions/files/ProductProfileRegistry.java:216]: Done.  Not sure why this happened.

File: src/main/java/org/candlepin/subscriptions/files/ProductUom.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    public ProductUom() {
+        // required for YAML
+    }
+
+    public ProductUom(String productId, String uom) {
+        this.productId = productId;
+        this.uom = uom;
+    }
+
+    @Override
```
Comment (dmckoske) [inline src/main/java/org/candlepin/subscriptions/files/ProductUom.java:50]: Should I remove this and hashCode, and add @EqualsAndHashCode and @ToString?

**Comment by dmckoske** [comment]
> After chatting with @kahowell I believe there is a part missing in this PR. Kevin mentioned that it was implied that ENT-3610 would include hooking the new configuration bits into appropriate place.
> 
> We need to get the metric value from the ProductProfile and replace the hard coded value here:
> 
> https://github.com/RedHatInsights/rhsm-subscriptions/blob/45f2786f45ccd316f1fa42188945dd34b34067f4/src/main/java/org/candlepin/subscriptions/files/ProductProfile.java#L75
> 
> productProfile.getMetricId() is called from the MarketplacePayloadMapper. You will need to make the connection.

I will look into this right away.

File: src/main/java/org/candlepin/subscriptions/files/ProductUom.java (hunk 7 above)
Comment (mstead) [reply src/main/java/org/candlepin/subscriptions/files/ProductUom.java:50]: Yes. In general, if there's nothing in those methods outside of the standard equals/hashcode pattern, then IMO it is safe to replace with the annotations.

**Comment by mstead** [comment]
> I will look into this right away.

Let me know if I can help in any way.

File: src/main/java/org/candlepin/subscriptions/files/ProductProfileRegistry.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (mstead) [reply src/main/java/org/candlepin/subscriptions/files/ProductProfileRegistry.java:216]: Still one reference remains on ```swatchProfileId``` parameter.

File: src/main/java/org/candlepin/subscriptions/files/ProductUom.java (hunk 7 above)
Comment (dmckoske) [reply src/main/java/org/candlepin/subscriptions/files/ProductUom.java:50]: Okay, good.  Done.

File: src/test/resources/test_product_profile_registry.yaml [hunk 8]
```diff
@@ -109,3 +109,8 @@ defaultSla: PREMIUM
 defaultUsage: PRODUCTION
 burstable: false
 architectureSwatchProductIdMap: {}
+marketplaceMetrics:
+  - metricId: redhat.com:openshiftdedicated:cpu_hour
+    uom: Cores
+    swatchProductIds:
+      - OpenShift-dedicated-metrics
```
Comment (kahowell) [inline src/test/resources/test_product_profile_registry.yaml:116]: Nitpick: empty line at EOF, please.

File: src/main/java/org/candlepin/subscriptions/marketplace/MarketplacePayloadMapper.java [hunk 9]
```diff
@@ -196,9 +196,11 @@ protected boolean isSnapshotPAYGEligible(TallySnapshot snapshot) {
             .forEach(measurement -> {
                 UsageMeasurement usageMeasurement = new UsageMeasurement();
                 usageMeasurement.setValue(measurement.getValue());
-                usageMeasurement.setMetricId(productProfile.getMetricId());
+                usageMeasurement.setMetricId(productProfile.metricByProductAndUom(productId,
+                    productProfile.fromTallyMeasurementUom(measurement.getUom())));
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/marketplace/MarketplacePayloadMapper.java:200]: I find this lookup here just a touch awkward. Can you please change the signature of metricByProductAndUom to accept `TallyMeasurement.Uom` (and just call `fromTallyMeasurementUom` internally)?

**Review by kahowell — CHANGES_REQUESTED** [review]
Few small requests for readability/tests, overall looking good.

File: src/main/java/org/candlepin/subscriptions/files/ProductProfile.java [hunk 10]
```diff
@@ -216,4 +165,17 @@ public boolean supportsGranularity(Granularity granularity) {
...
+        return metricByProductAndUom.get(new ProductUom(swatchProductId, uom.toString()));
+    }
+
+    public Measurement.Uom fromTallyMeasurementUom(TallyMeasurement.Uom uom) {
+        switch(uom) {
+            case SOCKETS:
+                return Measurement.Uom.SOCKETS;
+            case CORES:
+            default:
+                return Measurement.Uom.CORES;
```
[the remark this answers is not in the document]
Comment (dmckoske) [reply src/main/java/org/candlepin/subscriptions/files/ProductProfile.java:178]: Fair.  Done.

File: src/main/java/org/candlepin/subscriptions/files/ProductProfileRegistry.java [hunk 11]
```diff
@@ -53,6 +55,7 @@ public ProductProfileRegistry() {
         productProfilesByName = new HashMap<>();
         engProductIdToProfileMap = new HashMap<>();
         swatchProductIdToProfileMap = new HashMap<>();
+        metricByProductAndUom = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (dmckoske) [reply src/main/java/org/candlepin/subscriptions/files/ProductProfileRegistry.java:58]: Fair.  Done.

File: src/main/java/org/candlepin/subscriptions/marketplace/MarketplacePayloadMapper.java (hunk 9 above)
Comment (dmckoske) [reply src/main/java/org/candlepin/subscriptions/marketplace/MarketplacePayloadMapper.java:200]: Fair.  Done.

File: src/main/resources/product_profile_registry.yaml [hunk 12]
```diff
@@ -112,3 +112,8 @@ defaultSla: PREMIUM
 defaultUsage: PRODUCTION
 burstable: false
 architectureSwatchProductIdMap: {}
+marketplaceMetrics:
+  - metricId: redhat.com:openshiftdedicated:cpu_hour
+    uom: Cores
+    swatchProductIds:
+      - OpenShift-dedicated-metrics
```
[the remark this answers is not in the document]
Comment (dmckoske) [reply src/main/resources/product_profile_registry.yaml:119]: Yeah, Git flagged it too.  Done.

File: src/test/resources/test_product_profile_registry.yaml (hunk 8 above)
Comment (dmckoske) [reply src/test/resources/test_product_profile_registry.yaml:116]: Fair.  Done.

**Review by mstead — APPROVED** [review]

**Comment by mstead** [comment]
IMO looks good, just needs a final squash. @kahowell can merge when he is happy.

**Review by kahowell — APPROVED** [review]

**Pull request merged**